### PR TITLE
[GLib] New API to get the request body of WebKitURISchemeRequest

### PR DESCRIPTION
**Pull request by wusyong**
#### 5fe923aa166a3084c973020d99f34c9476dddece
<pre>
[GLib] New API to get the request body of WebKitURISchemeRequest
<a href="https://bugs.webkit.org/show_bug.cgi?id=252564">https://bugs.webkit.org/show_bug.cgi?id=252564</a>

Reviewed by Carlos Garcia Campos, Michael Catanzaro and Adrian Perez de Castro.

* Source/WebCore/platform/network/soup/ResourceRequest.h:
* Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp:
(WebCore::ResourceRequest::createInputStream const): Added new function
API
* Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp:
(webkit_uri_scheme_request_get_http_body): Added new function API
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp:
(testWebContextURIScheme): Expanded &apos;post&apos; URI scheme to test
webkit_uri_scheme_request_get_http_body

Canonical link: <a href="https://commits.webkit.org/260949@main">https://commits.webkit.org/260949@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/1d216eaca047b3959817b763dfacc921ca0ee54a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110030 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19130 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1470 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119064 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113982 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20597 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10323 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102287 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115776 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43555 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30206 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11832 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31544 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12454 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17821 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51159 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7590 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14249 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp
Comment (aperezdc) [inline Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp]: I would remove “http” from the documentation. URI scheme requests don't involve the HTTP protocol after all.

```suggestion
 * Get the request body.
 *
 * Returns: (transfer none): the body of the @request.
```

**Review by aperezdc — APPROVED** [review]
Thanks for the patch @wusyong, this looks neat 🏗️ 

We have just branched for 2.40 recently, but IMO this is quite nice to have and should be fine to add into the release branch if @carlosgcampos or @mcatanzaro agree&mdash;the patch is self-contained, only adds one public method, and comes with a test. WDYT?

(As with any API change, we need a second approval before landing, so let's wait and see what the other reviewers think.)

**Comment by aperezdc** [comment]
@wusyong I have changed the Bugzilla issue title to “[GLib] New API to get the request body of WebKitURISchemeRequest” because to better reflect what the feature is about, and also because it also bring the new API to the WPE port (GLib = GTK + WPE). Could you change the first line of the commit log accordingly? Thanks! 🙏🏼

**Comment by carlosgcampos** [comment]
Looks good to me, but I wonder if it would be easier to expose the body as a GBytes instead. In this case, the body is always serialized over IPC, so it fits in memory, and it's not expected to be processed as a stream, I guess.

File: Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp
Comment (mcatanzaro) [inline Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp:145]: Heh, so it is still serialized after all? We were previously serializing it twice prior to https://commits.webkit.org/260252@main? OK.

**Review by mcatanzaro — COMMENTED** [review]
Good work.

**Comment by mcatanzaro** [comment]
> Looks good to me, but I wonder if it would be easier to expose the body as a GBytes instead. In this case, the body is always serialized over IPC, so it fits in memory, and it's not expected to be processed as a stream, I guess.

GBytes is fine. Or simply `unsigned char *` or `uint8_t *`.

**Comment by wusyong** [comment]
I'll stick to GBytes since many APIs also return this type for bytes. I've updated the return type to GBytes. Please let me know if I missed anything.

File: Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp
Comment (mcatanzaro) [inline Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]:
```suggestion
GRefPtr<GBytes> ResourceRequest::createBodyBytes() const
```

File: Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp
Comment (mcatanzaro) [inline Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]:
```suggestion
    return adoptGRef(g_bytes_new_static(bytes.data(), bytes.size()));
```

File: Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp]: Oops:

```suggestion
```

File: Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp]: So we should always use GRefPtr internally. Then here at the outermost layer we can get rid of it:

```suggestion
    return request->priv->task->request().createBodyBytes().leakRef();
```

But it would also be OK to cache a copy like you had been doing with the GIOStream and return `(transfer none)`. I don't have a strong opinion on which way you choose, although I guess caching seems slightly nicer because one benefit of GBytes is that it's refcounted, and that benefit is lost if you create a copy each time. Not that we expect applications will likely be calling this multiple times per request.

File: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp
Comment (mcatanzaro) [inline Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp]: Making a note here just to ensure we don't forget to get rid of the GRefPtr here if you decide to cache the GBytes and return `(transfer none)`

**Review by mcatanzaro — APPROVED** [review]
By the way: it's very unusual for your first contribution to be new public API, so congrats on that. This is simple, adequately tested, and looks good.

File: Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp
Comment (wusyong) [reply Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp]: I changed to transfer full because I saw many APIs do so. I'll switch back to none.

File: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp
Comment (mcatanzaro) [inline Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp]: No need to use GRefPtr here anymore:

```suggestion
            GBytes* body = webkit_uri_scheme_request_get_http_body(request);
```

**Comment by mcatanzaro** [comment]
One last thing: please squash your two commits into one. We can't land pull requests with multiple commits for WebKit-specific reasons. So please fix that, and make the one-line change that I proposed above, and then this is good to go. And we'll backport to 2.40 as well so you can use it sooner rather than later.

File: Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp
Comment (mcatanzaro) [inline Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]: Ah, I was about to land this but realized this can actually be simplified further. There's no need to loop through resolvedFormData:

```suggestion
    auto bytes = formData->resolveBlobReferences()->flatten();
    if (!bytes.size())
        return nullptr;
    return adoptGRef(g_bytes_new_static(bytes.data(), bytes.size()));
```

Comment (mcatanzaro) [reply Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]: Another option would be even simpler:

```suggestion
    auto bytes = formData->resolveBlobReferences()->flatten();
    return adoptGRef(g_bytes_new_static(bytes.data(), bytes.size()));
```

In this case, the public API return value would be non-nullable, but the GBytes would be empty and contain NULL data of 0 size. Maybe this would be better so that callers don't have to NULL-check whether the GBytes itself is NULL? The caller would only need to check the data that it contains.

File: Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp]: Should either (a) change the implementation to ensure it does not return NULL, or (b) mark it as nullable:

```suggestion
 * Returns: (transfer none): (nullable): the body of the @request.
```

**Review by mcatanzaro — APPROVED** [review]

File: Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp
Comment (wusyong) [reply Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp]: formData could be empty, so I'll choose nullable.

File: Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp
Comment (carlosgcampos) [inline Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]: I don't think this is a good idea, flatten ignores blobs and files. I'm not sure if it would be possible to have a request body with blobs and files for custom uri schemes, though.

File: Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp
Comment (carlosgcampos) [inline Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]: We can't use `new_static` here, the Vector will be freed when returning from this function. We need to take a reference of FormData and use `new_with_free_func` to release the reference.

**Review by carlosgcampos — CHANGES_REQUESTED** [review]

File: Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp
Comment (carlosgcampos) [reply Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]: hmm, but if the form data can contain files... then the stream approach was better... I'm sorry I didn't realized

Comment (wusyong) [reply Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]: I hope it can be a stream type if that can happen. In my previous implementation, I used `webkitFormDataInputStreamNew`, which will get the type `WebKitFormDataInputStream`. This should behave like a stream, right? If you can agree, I'll fall back to the previous commit that returns input stream.

Comment (carlosgcampos) [reply Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]: Yes, sorry, I think the safest approach would be to return a stream like in your first patch

File: Source/WebCore/platform/network/soup/ResourceRequest.h
Comment (carlosgcampos) [inline Source/WebCore/platform/network/soup/ResourceRequest.h]: createBodyStream

File: Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp]: I don't think transfer none is useful here, the returned stream is not seekable, so after being read, if you call this again, you will get a stream that can't be read again.

File: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp
Comment (carlosgcampos) [inline Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp]: This works because the stream is pollable and it can read immediately, but you should use `g_input_stream_read_all` instead.

**Review by carlosgcampos — CHANGES_REQUESTED** [review]

File: Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp
Comment (wusyong) [reply Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp]: I'll change to transfer full and just return the stream to user then.

**Review by carlosgcampos — APPROVED** [review]

File: Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp
Comment (mcatanzaro) [inline Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]: This can still be simplified. We don't need to count up to the full length because it's not used:

```suggestion
    auto resolvedFormData = formData->resolveBlobReferences();
    for (auto& element : resolvedFormData->elements())
        if (element.lengthInBytes() > 0)
            return webkitFormDataInputStreamNew(WTFMove(resolvedFormData));
    return nullptr;
```

Comment (wusyong) [reply Source/WebCore/platform/network/soup/ResourceRequestSoup.cpp]: I'll need to add braces because `check-webkit-style` isn't happy with multi line control clauses without it.

**Pull request merged**
